skiko/src/awtMain/kotlin/org/jetbrains/skiko/swing/Direct3DSwingRedrawer.kt

```diff
@@ -0,0 +1,128 @@
+package org.jetbrains.skiko.swing
+
+import org.jetbrains.skia.*
+import org.jetbrains.skiko.*
+import java.awt.Graphics2D
+
+internal class Direct3DSwingRedrawer(
+    swingLayerProperties: SwingLayerProperties,
+    private val skikoView: SkikoView,
+    analytics: SkiaLayerAnalytics
+) : SwingRedrawerBase(swingLayerProperties, analytics, GraphicsApi.DIRECT3D) {
+    companion object {
+        init {
+            Library.load()
+        }
+    }
+
+    private val adapter = chooseAdapter(swingLayerProperties.adapterPriority.ordinal).also {
+        onDeviceChosen("DirectX12") // TODO: properly get name
+    }
+
+    private val device = createDirectXOffscreenDevice(adapter)
+
+    private val swingOffscreenDrawer = SwingOffscreenDrawer(swingLayerProperties)
+
+    private val context = DirectContext(
+        makeDirectXContext(device)
+    )
+
+    private var texturePtr: Long = 0
+    private var renderTarget: BackendRenderTarget? = null
+    private var surface: Surface? = null
+
+    private var bytesToDraw = ByteArray(0)
+    private val rowBytesAlignment = getAlignment().toInt()
+    private val widthSizeAlignment = rowBytesAlignment / 4
+
+    init {
+        onContextInit()
+    }
+
+    override fun dispose() {
+        surface?.close()
+        renderTarget?.close()
+        bytesToDraw = ByteArray(0)
+        context.close()
+        disposeDevice(device)
+        super.dispose()
+    }
+
+    override fun onRender(g: Graphics2D, width: Int, height: Int, nanoTime: Long) {
+        val alignedWidth = if (width % widthSizeAlignment != 0) {
+            width + widthSizeAlignment - (width % widthSizeAlignment);
+        } else {
+            width
+        }
+
+        val newTexturePtr = getRenderTargetTexture(device, alignedWidth, height)
+
+        if (newTexturePtr == 0.toLong()) {
+            throw RenderException("Can't allocate DirectX resources")
+        }
+
+        if (newTexturePtr != texturePtr) {
+            texturePtr = newTexturePtr
```

Comment:
We should close renderTarget and surface in this case.
BTW can we create surface and renderTarget on each render as we do with Metal? Does it significantly affect performance?

Reply:
Yes, indeed, they leak now on each resize.  Recreating those makes sense when their wrapping entity or usage options change, which is not the case. I don't think it affects performance since those are lightweight logical entities (assuming Skia doesn't do something dumb or too smart because of this), but why would we opt-in into that way?